src/jllvm/support/Encoding.cpp

```diff
@@ -52,3 +52,28 @@ std::pair<std::vector<std::uint8_t>, jllvm::CompactEncoding> jllvm::toJavaCompac
     }
     return {std::move(result), CompactEncoding::Utf16BE};
 }
+
+std::string jllvm::fromJavaCompactEncoding(std::pair<llvm::ArrayRef<std::uint8_t>, CompactEncoding> compactEncoding)
+{
+    auto [buffer, encoding] = compactEncoding;
+    if (encoding == CompactEncoding::Latin1)
+    {
+        return std::string{buffer.begin(), buffer.end()};
+    }
+
+    std::vector<char> utf16{buffer.begin(), buffer.end()};
+    std::string string;
+
+    if constexpr (llvm::support::endianness::native != llvm::support::endianness::big)
+    {
+        assert(utf16.size() % 2 == 0);
+        for (auto it = utf16.begin(); it != utf16.end();)
+        {
+            std::swap(*it++, *it++);
+        }
+    }
+
+    assert(llvm::convertUTF16ToUTF8String(utf16, string));
```

Comment:
You have to be careful here, the contents of the `assert` expression is is a NOOP in release builds. Just move the call into out of the assert and check its return value explicitly